File: 2024/Julia/day19.jl
using Memoize

input = "../R/inst/input19.txt"
p = split(read(input, String), r"\R\R")
towels = split(p[1], ", ")
patterns = split(p[2], r"\n")

substr(s, x) = last(s, length(s) - length(x))

@memoize function count_designs(p) 
    if length(p) == 0 return 1 end
    opts = filter(x -> startswith(p, x), towels)
    if length(opts) == 0 return 0 end
    return sum(map(x -> count_designs(substr(p, x)), opts))
end

solves = map(count_designs, patterns)
println(sum(solves .> 0)) # Part 1
println(sum(solves))      # Part 2

File: 2024/Python/day19.py
from functools import cache

@cache
def count_designs(p):
    if len(p) == 0:
        return 1
    opts = [x for x in towels if p[0:len(x)] == x]
    if len(opts) == 0:
        return 0
    return sum([count_designs(p[len(x):]) for x in opts])

with open("../R/inst/input19.txt") as f:
    input = f.read().split("\n")

towels = input[0].split(", ")
patterns = [x for x in input[2:] if x != ""]

solves = [count_designs(p) for p in patterns]
print(len([x for x in solves if x > 0])) # Part 1
print(sum(solves))                       # Part 2
